packages/melody-component/src/hookComponent.js

```diff
@@ -0,0 +1,223 @@
+    const dataPrev = hooks[hooksPointer][2];
+    const dirty =
+        shouldUpdateOrDataArray === true ||
+        (shouldUpdateOrDataArray &&
+            shouldUpdateOrDataArray.length &&
+            !shallowEqualsArray(dataPrev, shouldUpdateOrDataArray));
+
+    hooks[hooksPointer][2] = shouldUpdateOrDataArray;
+    hooks[hooksPointer][3] = dirty;
+};
+
+function Component(element, componentFn) {
+    // part of the public API
+    this.el = element;
+    this.refs = Object.create(null);
+
+    // needed for this type of component
+    this.componentFn = componentFn;
+    this.hooks = [];
+    this.hooksPointer = -1;
+    this.isMounted = false;
+    this.props = null;
+    this.state = {};
+    this.data = null;
+
+    this.setState = this.setState.bind(this);
+}
+
+Object.assign(Component.prototype, {
+    /**
+     * Set new properties for the Component.
+     * This might cause the component to request an update.
+     */
+    apply(props) {
+        if (shallowEquals(props, this.props)) {
+            return;
+        }
+        this.props = props || null;
+        this.enqueueComponent();
+    },
+
+    /**
+     * Executed after a component has been mounted or updated.
+     * After this method has been triggered, the component is considered stable and
+     * accessing the DOM should be safe.
+     * The children of this Component might not have been rendered.
+     */
+    notify() {
+        const hooks = this.hooks;
+        for (let i = 0, l = hooks.length; i < l; i++) {
+            const hook = hooks[i];
+            const type = hook[0];
+
+            if (type !== HOOK_TYPE_USE_EFFECT) continue;
+
+            const dirty = hook[3];
+            if (!dirty) continue;
+
+            const callback = hook[1];
+            const unsubscribe = hook[4];
+
+            if (unsubscribe) unsubscribe();
+            const unsubscribeNext = callback() || null;
+            hook[4] = unsubscribeNext;
+        }
+
+        if (this.isMounted) {
+            if (options.afterUpdate) {
+                options.afterUpdate(this);
+            }
+        } else {
+            this.isMounted = true;
+            if (options.afterMount) {
+                options.afterMount(this);
+            }
+        }
+    },
+
+    setState(hookIndex, valueNext) {
+        const value = this.state[hookIndex];
+        if (shallowEqualsScalar(value, valueNext)) {
+            return;
+        }
+        this.state[hookIndex] = valueNext;
+        this.enqueueComponent();
+    },
+
+    enqueueComponent() {
+        if (currentComponent) {
+            throw new Error('Cannot compose multiple components at a time');
+        }
+        currentComponent = this;
+        this.hooksPointer = -1;
+        this.data = this.componentFn(this.props) || {};
+        currentComponent = null;
+
+        if (this.el) {
+            enqueueComponent(this);
+        }
+    },
+
+    /**
+     * Invoked when a component should render itself.
+     */
+    render() {},
```

Comment:
we can consider unsubscribe effect after each render

Reply:
what do you mean?